# Healthcare Analytics Data Warehouse & Executive Reporting Platform

## Project Objectives

This project develops a healthcare analytics platform using CMS Medicare Provider Utilization data.

### Business Goals

- Analyze provider utilization and reimbursement trends
- Evaluate provider performance
- Monitor beneficiary demographics and health characteristics
- Build a dimensional data warehouse for analytical reporting
- Develop executive dashboards for healthcare decision-makers

## Environment Configuration

Import required libraries and configure notebook display settings.

In [6]:
import pandas as pd
import numpy as np
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)
print("Pandas Version:", pd.__version__)
print("NumPy Version:", np.__version__)

Pandas Version: 3.0.3
NumPy Version: 2.4.6


## Data Acquisition

Load the CMS Medicare Physician & Other Practitioners dataset from the raw data layer.

In [7]:
claims = pd.read_csv("../data/raw/claims.csv", low_memory=False)
print(f"Rows: {claims.shape[0]:,}")
print(f"Columns: {claims.shape[1]:,}")

Rows: 1,296,739
Columns: 81


## Dataset Preview

Review the raw dataset structure and validate successful ingestion.

In [8]:
claims.head()

,Rndrng_NPI,Rndrng_Prvdr_Last_Org_Name,Rndrng_Prvdr_First_Name,Rndrng_Prvdr_MI,Rndrng_Prvdr_Crdntls,Rndrng_Prvdr_Ent_Cd,Rndrng_Prvdr_St1,Rndrng_Prvdr_St2,Rndrng_Prvdr_City,Rndrng_Prvdr_State_Abrvtn,Rndrng_Prvdr_State_FIPS,Rndrng_Prvdr_Zip5,Rndrng_Prvdr_RUCA,Rndrng_Prvdr_RUCA_Desc,Rndrng_Prvdr_Cntry,Rndrng_Prvdr_Type,Rndrng_Prvdr_Mdcr_Prtcptg_Ind,Tot_HCPCS_Cds,Tot_Benes,Tot_Srvcs,Tot_Sbmtd_Chrg,Tot_Mdcr_Alowd_Amt,Tot_Mdcr_Pymt_Amt,Tot_Mdcr_Stdzd_Amt,Drug_Sprsn_Ind,Drug_Tot_HCPCS_Cds,Drug_Tot_Benes,Drug_Tot_Srvcs,Drug_Sbmtd_Chrg,Drug_Mdcr_Alowd_Amt,Drug_Mdcr_Pymt_Amt,Drug_Mdcr_Stdzd_Amt,Med_Sprsn_Ind,Med_Tot_HCPCS_Cds,Med_Tot_Benes,Med_Tot_Srvcs,Med_Sbmtd_Chrg,Med_Mdcr_Alowd_Amt,Med_Mdcr_Pymt_Amt,Med_Mdcr_Stdzd_Amt,Bene_Avg_Age,Bene_Age_LT_65_Cnt,Bene_Age_65_74_Cnt,Bene_Age_75_84_Cnt,Bene_Age_GT_84_Cnt,Bene_Feml_Cnt,Bene_Male_Cnt,Bene_Race_Wht_Cnt,Bene_Race_Black_Cnt,Bene_Race_API_Cnt,Bene_Race_Hspnc_Cnt,Bene_Race_NatInd_Cnt,Bene_Race_Othr_Cnt,Bene_Dual_Cnt,Bene_Ndual_Cnt,Bene_CC_BH_ADHD_OthCD_V1_Pct,Bene_CC_BH_Alcohol_Drug_V1_Pct,Bene_CC_BH_Tobacco_V1_Pct,Bene_CC_BH_Alz_NonAlzdem_V2_Pct,Bene_CC_BH_Anxiety_V1_Pct,Bene_CC_BH_Bipolar_V1_Pct,Bene_CC_BH_Mood_V2_Pct,Bene_CC_BH_Depress_V1_Pct,Bene_CC_BH_PD_V1_Pct,Bene_CC_BH_PTSD_V1_Pct,Bene_CC_BH_Schizo_OthPsy_V1_Pct,Bene_CC_PH_Asthma_V2_Pct,Bene_CC_PH_Afib_V2_Pct,Bene_CC_PH_Cancer6_V2_Pct,Bene_CC_PH_CKD_V2_Pct,Bene_CC_PH_COPD_V2_Pct,Bene_CC_PH_Diabetes_V2_Pct,Bene_CC_PH_HF_NonIHD_V2_Pct,Bene_CC_PH_Hyperlipidemia_V2_Pct,Bene_CC_PH_Hypertension_V2_Pct,Bene_CC_PH_IschemicHeart_V2_Pct,Bene_CC_PH_Osteoporosis_V2_Pct,Bene_CC_PH_Parkinson_V2_Pct,Bene_CC_PH_Arthritis_V2_Pct,Bene_CC_PH_Stroke_TIA_V2_Pct,Bene_Avg_Risk_Scre
0,1003000126,Enkeshafi,Ardalan,NaN,M.D.,I,6410 Rockledge Dr Ste 304,NaN,Bethesda,MD,24,20817,1.0,Metropolitan area core: primary flow within an...,US,Internal Medicine,Y,15,328,399.0,202783.88,46472.65,35325.46,34781.01,NaN,0.0,0.0,0.0,0.0,0.00,0.0,0.0,NaN,15.0,328.0,399.0,202783.88,46472.65,35325.46,34781.01,76,37.0,94.0,130.0,67.0,171.0,157.0,281.0,21.0,NaN,NaN,NaN,NaN,77.0,251.0,NaN,15.0,19.0,27.0,46.0,9.0,52.0,48.0,NaN,NaN,5.0,17.0,39.0,27.0,55.0,38.0,48.0,48.0,75.0,75.0,51.0,18.0,4.0,63.0,21.0,2.5660
1,1003000134,Cibull,Thomas,L,M.D.,I,2650 Ridge Ave,Evanston Hospital,Evanston,IL,17,60201,1.0,Metropolitan area core: primary flow within an...,US,Pathology,Y,21,912,2023.0,302581.00,69422.59,50178.11,47710.61,NaN,0.0,0.0,0.0,0.0,0.00,0.0,0.0,NaN,21.0,912.0,2023.0,302581.00,69422.59,50178.11,47710.61,76,16.0,419.0,337.0,140.0,455.0,457.0,806.0,13.0,29.0,23.0,0.0,41.0,48.0,864.0,NaN,3.0,4.0,5.0,16.0,3.0,20.0,17.0,1.0,NaN,NaN,6.0,17.0,17.0,21.0,10.0,23.0,14.0,75.0,64.0,26.0,15.0,2.0,47.0,6.0,1.1958
2,1003000142,Khalil,Rashid,NaN,M.D.,I,4126 N Holland Sylvania Rd,Suite 220,Toledo,OH,39,43623,1.0,Metropolitan area core: primary flow within an...,US,Anesthesiology,Y,41,316,1369.0,381301.00,133096.84,101577.94,103554.75,NaN,2.0,20.0,261.0,4154.0,29.87,23.8,23.8,NaN,39.0,316.0,1108.0,377147.00,133066.97,101554.14,103530.95,70,84.0,127.0,84.0,21.0,192.0,124.0,199.0,97.0,NaN,NaN,0.0,NaN,112.0,204.0,NaN,20.0,21.0,7.0,53.0,7.0,46.0,44.0,NaN,4.0,5.0,15.0,17.0,12.0,34.0,32.0,45.0,27.0,75.0,75.0,36.0,17.0,NaN,75.0,11.0,1.6931
3,1003000423,Velotta,Jennifer,A,M.D.,I,8300 Tyler Blvd Ste 300,NaN,Mentor,OH,39,44060,1.0,Metropolitan area core: primary flow within an...,US,Obstetrics & Gynecology,Y,18,75,140.0,22015.00,8708.74,7134.32,7502.21,NaN,0.0,0.0,0.0,0.0,0.00,0.0,0.0,NaN,18.0,75.0,140.0,22015.00,8708.74,7134.32,7502.21,70,NaN,50.0,NaN,NaN,75.0,0.0,NaN,NaN,NaN,NaN,NaN,NaN,11.0,64.0,NaN,NaN,NaN,NaN,35.0,NaN,29.0,28.0,NaN,NaN,NaN,NaN,NaN,20.0,NaN,NaN,17.0,NaN,73.0,71.0,17.0,25.0,NaN,45.0,NaN,0.7931
4,1003000480,Rothchild,Kevin,B,MD,I,12605 E 16th Ave,NaN,Aurora,CO,08,80045,1.0,Metropolitan area core: primary flow within an...,US,General Surgery,Y,28,101,136.0,120420.00,24416.29,18766.80,18618.81,NaN,0.0,0.0,0.0,0.0,0.00,0.0,0.0,NaN,28.0,101.0,136.0,120420.00,24416.29,18766.80,18618.81,68,N

## Data Understanding

The dataset contains provider-level healthcare utilization, reimbursement, demographic, and risk metrics reported by CMS.

### Analytical Grain

Each record represents a unique healthcare provider identified by National Provider Identifier (NPI) for a reporting year.

### Analysis Domains

- Provider Performance
- Healthcare Utilization
- Medicare Reimbursements
- Beneficiary Demographics
- Chronic Condition Prevalence
- Geographic Analysis

In [9]:
claims.info()

<class 'pandas.DataFrame'>
RangeIndex: 1296739 entries, 0 to 1296738
Data columns (total 81 columns):
 #   Column                            Non-Null Count    Dtype  
---  ------                            --------------    -----  
 0   Rndrng_NPI                        1296739 non-null  int64  
 1   Rndrng_Prvdr_Last_Org_Name        1296739 non-null  str    
 2   Rndrng_Prvdr_First_Name           1235756 non-null  str    
 3   Rndrng_Prvdr_MI                   810051 non-null   str    
 4   Rndrng_Prvdr_Crdntls              1123375 non-null  str    
 5   Rndrng_Prvdr_Ent_Cd               1296739 non-null  str    
 6   Rndrng_Prvdr_St1                  1296739 non-null  str    
 7   Rndrng_Prvdr_St2                  243316 non-null   str    
 8   Rndrng_Prvdr_City                 1296739 non-null  str    
 9   Rndrng_Prvdr_State_Abrvtn         1296739 non-null  str    
 10  Rndrng_Prvdr_State_FIPS           1296738 non-null  str    
 11  Rndrng_Prvdr_Zip5                 1296738 non-nu

## Data Quality Assessment

This section evaluates dataset structure, completeness, and potential data quality issues prior to dimensional modeling and ETL development.

### Assessment Areas

- Dataset Size
- Data Types
- Missing Values
- Key Fields
- Candidate Dimensions
- Candidate Measures

## Key Analysis

Validate the uniqueness of the National Provider Identifier (NPI) and confirm dataset grain.

In [11]:
total_rows = len(claims)
unique_npis = claims["Rndrng_NPI"].nunique()
print(f"Total Rows      : {total_rows:,}")
print(f"Unique NPIs     : {unique_npis:,}")
print(f"Duplicate NPIs  : {total_rows - unique_npis:,}")

Total Rows      : 1,296,739
Unique NPIs     : 1,296,739
Duplicate NPIs  : 0


## Missing Value Analysis

Evaluate data completeness across all attributes and identify fields requiring imputation, suppression handling, or exclusion from analytical models.

In [13]:
missing_values = (claims.isnull().sum().reset_index())
missing_values.columns = ["column_name", "missing_count"]
missing_values["missing_pct"] = ( missing_values["missing_count"] / len(claims) * 100).round(2)
missing_values = (missing_values.sort_values("missing_pct", ascending=False))
missing_values.head(25)

,column_name,missing_count,missing_pct
24,Drug_Sprsn_Ind,1145043,88.30
32,Med_Sprsn_Ind,1145043,88.30
49,Bene_Race_API_Cnt,1062540,81.94
7,Rndrng_Prvdr_St2,1053423,81.24
52,Bene_Race_Othr_Cnt,1052600,81.17
50,Bene_Race_Hspnc_Cnt,988886,76.26
48,Bene_Race_Black_Cnt,924899,71.32
64,Bene_CC_BH_PTSD_V1_Pct,882169,68.03
55,Bene_CC_BH_ADHD_OthCD_V1_Pct,875024,67.48
63,Bene_CC_BH_PD_V1_Pct,852076,65.71


## Missing Value Interpretation

Healthcare datasets frequently contain suppressed values to protect beneficiary privacy.

According to CMS reporting rules, metrics based on fewer than 11 beneficiaries may be suppressed and represented as null values.

Therefore, null values in this dataset do not always indicate data quality issues and must be interpreted within the healthcare reporting context.

## Preliminary Data Quality Decisions

| Category | Action |
|-----------|----------|
| Address Line 2 | Retain in raw layer, exclude from analytics |
| Suppression Indicators | Preserve as-is |
| Demographic Counts | Convert null values to 0 in curated layer |
| Chronic Condition Percentages | Preserve null values until feature engineering |
| Provider Name Fields | Retain |
| Financial Metrics | Retain |
| Geographic Fields | Retain |

## Duplicate Record Analysis

Validate whether duplicate rows exist within the source dataset.

In [14]:
duplicate_rows = claims.duplicated().sum()
print(f"Duplicate Rows: {duplicate_rows:,}")

Duplicate Rows: 0


## Dimensional Modeling

A dimensional model is designed to support analytical reporting and executive dashboarding.

The source dataset contains provider-level healthcare utilization and reimbursement metrics.

Based on the validated grain, a star schema will be implemented consisting of:

### Fact Table

- fact_provider_utilization

### Dimension Tables

- dim_provider
- dim_location
- dim_beneficiary_profile
- dim_date

In [16]:
provider_cols = [col for col in claims.columns
    if "Rndrng_Prvdr" in col]
provider_cols

['Rndrng_Prvdr_Last_Org_Name',
 'Rndrng_Prvdr_First_Name',
 'Rndrng_Prvdr_MI',
 'Rndrng_Prvdr_Crdntls',
 'Rndrng_Prvdr_Ent_Cd',
 'Rndrng_Prvdr_St1',
 'Rndrng_Prvdr_St2',
 'Rndrng_Prvdr_City',
 'Rndrng_Prvdr_State_Abrvtn',
 'Rndrng_Prvdr_State_FIPS',
 'Rndrng_Prvdr_Zip5',
 'Rndrng_Prvdr_RUCA',
 'Rndrng_Prvdr_RUCA_Desc',
 'Rndrng_Prvdr_Cntry',
 'Rndrng_Prvdr_Type',
 'Rndrng_Prvdr_Mdcr_Prtcptg_Ind']

## Provider Dimension Analysis

The provider dimension contains descriptive attributes related to healthcare providers.

These attributes are used for slicing, filtering, grouping, and reporting across analytical measures stored in the fact table.

## Provider Type Distribution

Analyze provider specialties represented within the dataset.

In [17]:
claims["Rndrng_Prvdr_Type"].nunique()

113

## Provider Specialty Profile

Analyze provider specialty distribution to understand the healthcare provider landscape represented in the dataset.

In [18]:
claims["Rndrng_Prvdr_Type"] \
    .value_counts() \
    .head(20)

Rndrng_Prvdr_Type
Nurse Practitioner                               203212
Physician Assistant                              112514
Internal Medicine                                 93627
Family Practice                                   83639
Physical Therapist in Private Practice            77901
Certified Registered Nurse Anesthetist (CRNA)     53495
Emergency Medicine                                50101
Anesthesiology                                    41082
Diagnostic Radiology                              32843
Chiropractic                                      32394
Optometry                                         31979
Mass Immunizer Roster Biller                      26738
Obstetrics & Gynecology                           22815
Orthopedic Surgery                                21674
General Surgery                                   21065
Hospitalist                                       20529
Psychiatry                                        19778
Cardiology                    

## Provider Specialty Analysis

The dataset contains 113 unique provider specialties.

The provider population is dominated by advanced practice providers, primary care providers, and high-volume clinical specialties.

This distribution will support provider performance benchmarking, reimbursement analysis, and utilization reporting.

## Provider Dimension Design

The provider dimension stores descriptive attributes related to healthcare providers.

### Business Purpose

Supports:

- Provider performance analysis
- Provider benchmarking
- Specialty reporting
- Reimbursement analysis
- Executive healthcare reporting

In [19]:
provider_cols = [
    "Rndrng_NPI",
    "Rndrng_Prvdr_Last_Org_Name",
    "Rndrng_Prvdr_First_Name",
    "Rndrng_Prvdr_MI",
    "Rndrng_Prvdr_Crdntls",
    "Rndrng_Prvdr_Ent_Cd",
    "Rndrng_Prvdr_Type",
    "Rndrng_Prvdr_Mdcr_Prtcptg_Ind" ]
dim_provider = claims[provider_cols].copy()
dim_provider.head()

,Rndrng_NPI,Rndrng_Prvdr_Last_Org_Name,Rndrng_Prvdr_First_Name,Rndrng_Prvdr_MI,Rndrng_Prvdr_Crdntls,Rndrng_Prvdr_Ent_Cd,Rndrng_Prvdr_Type,Rndrng_Prvdr_Mdcr_Prtcptg_Ind
0,1003000126,Enkeshafi,Ardalan,NaN,M.D.,I,Internal Medicine,Y
1,1003000134,Cibull,Thomas,L,M.D.,I,Pathology,Y
2,1003000142,Khalil,Rashid,NaN,M.D.,I,Anesthesiology,Y
3,1003000423,Velotta,Jennifer,A,M.D.,I,Obstetrics & Gynecology,Y
4,1003000480,Rothchild,Kevin,B,MD,I,General Surgery,Y


In [22]:
dim_provider.insert(0, "provider_key",
    range(1, len(dim_provider) + 1)
)
dim_provider.head()

,provider_key,provider_npi,provider_last_org_name,provider_first_name,provider_middle_initial,provider_credentials,entity_code,provider_type,medicare_participation_flag
0,1,1003000126,Enkeshafi,Ardalan,NaN,M.D.,I,Internal Medicine,Y
1,2,1003000134,Cibull,Thomas,L,M.D.,I,Pathology,Y
2,3,1003000142,Khalil,Rashid,NaN,M.D.,I,Anesthesiology,Y
3,4,1003000423,Velotta,Jennifer,A,M.D.,I,Obstetrics & Gynecology,Y
4,5,1003000480,Rothchild,Kevin,B,MD,I,General Surgery,Y


## Provider Dimension Standardization

The provider dimension is standardized using business-friendly column names to improve readability, maintainability, and reporting usability.

In [23]:
dim_provider = dim_provider.rename(
    columns={
        "Rndrng_NPI": "provider_npi",
        "Rndrng_Prvdr_Last_Org_Name": "provider_last_org_name",
        "Rndrng_Prvdr_First_Name": "provider_first_name",
        "Rndrng_Prvdr_MI": "provider_middle_initial",
        "Rndrng_Prvdr_Crdntls": "provider_credentials",
        "Rndrng_Prvdr_Ent_Cd": "entity_code",
        "Rndrng_Prvdr_Type": "provider_type",
        "Rndrng_Prvdr_Mdcr_Prtcptg_Ind": "medicare_participation_flag"
    }
)
dim_provider.head()

,provider_key,provider_npi,provider_last_org_name,provider_first_name,provider_middle_initial,provider_credentials,entity_code,provider_type,medicare_participation_flag
0,1,1003000126,Enkeshafi,Ardalan,NaN,M.D.,I,Internal Medicine,Y
1,2,1003000134,Cibull,Thomas,L,M.D.,I,Pathology,Y
2,3,1003000142,Khalil,Rashid,NaN,M.D.,I,Anesthesiology,Y
3,4,1003000423,Velotta,Jennifer,A,M.D.,I,Obstetrics & Gynecology,Y
4,5,1003000480,Rothchild,Kevin,B,MD,I,General Surgery,Y


In [24]:
dim_provider.columns.tolist()

['provider_key',
 'provider_npi',
 'provider_last_org_name',
 'provider_first_name',
 'provider_middle_initial',
 'provider_credentials',
 'entity_code',
 'provider_type',
 'medicare_participation_flag']

## Location Dimension Analysis

The location dimension contains geographic attributes associated with healthcare providers.

This dimension will support:

- Geographic utilization analysis
- State-level reimbursement analysis
- Rural vs Urban reporting
- Regional healthcare access analysis

In [25]:
location_cols = [
    "Rndrng_Prvdr_City",
    "Rndrng_Prvdr_State_Abrvtn",
    "Rndrng_Prvdr_State_FIPS",
    "Rndrng_Prvdr_Zip5",
    "Rndrng_Prvdr_RUCA",
    "Rndrng_Prvdr_RUCA_Desc",
    "Rndrng_Prvdr_Cntry"
]
claims[location_cols].head()

,Rndrng_Prvdr_City,Rndrng_Prvdr_State_Abrvtn,Rndrng_Prvdr_State_FIPS,Rndrng_Prvdr_Zip5,Rndrng_Prvdr_RUCA,Rndrng_Prvdr_RUCA_Desc,Rndrng_Prvdr_Cntry
0,Bethesda,MD,24,20817,1.0,Metropolitan area core: primary flow within an...,US
1,Evanston,IL,17,60201,1.0,Metropolitan area core: primary flow within an...,US
2,Toledo,OH,39,43623,1.0,Metropolitan area core: primary flow within an...,US
3,Mentor,OH,39,44060,1.0,Metropolitan area core: primary flow within an...,US
4,Aurora,CO,08,80045,1.0,Metropolitan area core: primary flow within an...,US


In [26]:
claims[location_cols] \
    .drop_duplicates() \
    .shape

(26836, 7)

## Location Dimension Construction

The location dimension stores geographic attributes associated with healthcare providers.

This dimension supports:

- State-level reporting
- Geographic utilization analysis
- Rural versus urban comparisons
- Regional reimbursement analysis

In [27]:
dim_location = (
    claims[location_cols]
    .drop_duplicates()
    .copy()
)
dim_location.head()

,Rndrng_Prvdr_City,Rndrng_Prvdr_State_Abrvtn,Rndrng_Prvdr_State_FIPS,Rndrng_Prvdr_Zip5,Rndrng_Prvdr_RUCA,Rndrng_Prvdr_RUCA_Desc,Rndrng_Prvdr_Cntry
0,Bethesda,MD,24,20817,1.0,Metropolitan area core: primary flow within an...,US
1,Evanston,IL,17,60201,1.0,Metropolitan area core: primary flow within an...,US
2,Toledo,OH,39,43623,1.0,Metropolitan area core: primary flow within an...,US
3,Mentor,OH,39,44060,1.0,Metropolitan area core: primary flow within an...,US
4,Aurora,CO,08,80045,1.0,Metropolitan area core: primary flow within an...,US


In [29]:
dim_location = dim_location.rename(
    columns={
        "Rndrng_Prvdr_City": "city",
        "Rndrng_Prvdr_State_Abrvtn": "state_abbreviation",
        "Rndrng_Prvdr_State_FIPS": "state_fips",
        "Rndrng_Prvdr_Zip5": "zip_code",
        "Rndrng_Prvdr_RUCA": "ruca_code",
        "Rndrng_Prvdr_RUCA_Desc": "ruca_description",
        "Rndrng_Prvdr_Cntry": "country"
    }
)
dim_location.head()

,city,state_abbreviation,state_fips,zip_code,ruca_code,ruca_description,country
0,Bethesda,MD,24,20817,1.0,Metropolitan area core: primary flow within an...,US
1,Evanston,IL,17,60201,1.0,Metropolitan area core: primary flow within an...,US
2,Toledo,OH,39,43623,1.0,Metropolitan area core: primary flow within an...,US
3,Mentor,OH,39,44060,1.0,Metropolitan area core: primary flow within an...,US
4,Aurora,CO,08,80045,1.0,Metropolitan area core: primary flow within an...,US


In [30]:
dim_location.insert(0, "location_key",
    range(1, len(dim_location) + 1)
)
dim_location.head()

,location_key,city,state_abbreviation,state_fips,zip_code,ruca_code,ruca_description,country
0,1,Bethesda,MD,24,20817,1.0,Metropolitan area core: primary flow within an...,US
1,2,Evanston,IL,17,60201,1.0,Metropolitan area core: primary flow within an...,US
2,3,Toledo,OH,39,43623,1.0,Metropolitan area core: primary flow within an...,US
3,4,Mentor,OH,39,44060,1.0,Metropolitan area core: primary flow within an...,US
4,5,Aurora,CO,08,80045,1.0,Metropolitan area core: primary flow within an...,US


## Location Dimension Validation

Validate uniqueness and surrogate key generation.

In [31]:
print(f"Location Rows : {len(dim_location):,}")
print(f"Unique Keys   : {dim_location['location_key'].nunique():,}")

Location Rows : 26,836
Unique Keys   : 26,836


## Fact Table Design

The fact table stores provider-level utilization, reimbursement, demographic, and risk metrics.

Each record represents a unique healthcare provider for the reporting year.

### Fact Grain

One record per provider (NPI) per reporting year.

In [33]:
measure_cols = [col
    for col in claims.columns
    if col not in [
        "Rndrng_NPI",
        "Rndrng_Prvdr_Last_Org_Name",
        "Rndrng_Prvdr_First_Name",
        "Rndrng_Prvdr_MI",
        "Rndrng_Prvdr_Crdntls",
        "Rndrng_Prvdr_Ent_Cd",
        "Rndrng_Prvdr_St1",
        "Rndrng_Prvdr_St2",
        "Rndrng_Prvdr_City",
        "Rndrng_Prvdr_State_Abrvtn",
        "Rndrng_Prvdr_State_FIPS",
        "Rndrng_Prvdr_Zip5",
        "Rndrng_Prvdr_RUCA",
        "Rndrng_Prvdr_RUCA_Desc",
        "Rndrng_Prvdr_Cntry",
        "Rndrng_Prvdr_Type",
        "Rndrng_Prvdr_Mdcr_Prtcptg_Ind"
    ]
]
len(measure_cols)

64

## Date Dimension Design

The date dimension provides a standardized time structure for analytical reporting.

Although the current dataset contains a single reporting year, a date dimension is implemented to support future historical data integration and time-series analysis.

In [34]:
dim_date = pd.DataFrame(
    {
        "date_key": [20240101],
        "reporting_year": [2024],
        "quarter": ["Annual"],
        "month": ["Annual"]
    }
)
dim_date

,date_key,reporting_year,quarter,month
0,20240101,2024,Annual,Annual


In [35]:
dim_date.info()

<class 'pandas.DataFrame'>
RangeIndex: 1 entries, 0 to 0
Data columns (total 4 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   date_key        1 non-null      int64
 1   reporting_year  1 non-null      int64
 2   quarter         1 non-null      str  
 3   month           1 non-null      str  
dtypes: int64(2), str(2)
memory usage: 164.0 bytes


## Fact Table Preparation

Before constructing the fact table, surrogate keys from the dimensions are mapped back to the source dataset.

This process establishes relationships between dimensions and the fact table and completes the star schema structure.

In [36]:
location_lookup = dim_location[
    [
        "location_key",
        "city",
        "state_abbreviation",
        "state_fips",
        "zip_code",
        "ruca_code",
        "ruca_description",
        "country"
    ]
]

In [38]:
claims_fact = claims.copy()
claims_fact = claims_fact.merge(
    location_lookup,
    left_on=[
        "Rndrng_Prvdr_City",
        "Rndrng_Prvdr_State_Abrvtn",
        "Rndrng_Prvdr_State_FIPS",
        "Rndrng_Prvdr_Zip5",
        "Rndrng_Prvdr_RUCA",
        "Rndrng_Prvdr_RUCA_Desc",
        "Rndrng_Prvdr_Cntry"
    ],
    right_on=[
        "city",
        "state_abbreviation",
        "state_fips",
        "zip_code",
        "ruca_code",
        "ruca_description",
        "country"
    ],
    how="left"
)
claims_fact.shape

(1296739, 89)

In [39]:
claims_fact["location_key"].isnull().sum()

np.int64(0)

## Provider Key Mapping

Map provider surrogate keys into the analytical fact table to establish relationships between the fact table and provider dimension.

In [40]:
provider_lookup = dim_provider[
    [ "provider_key", "provider_npi" ]
]

In [41]:
claims_fact = claims_fact.merge(
    provider_lookup,
    left_on="Rndrng_NPI",
    right_on="provider_npi",
    how="left"
)
claims_fact.shape

(1296739, 91)

In [42]:
claims_fact["provider_key"].isnull().sum()

np.int64(0)

In [43]:
claims_fact["date_key"] = 20240101

In [44]:
claims_fact[
    [
        "provider_key",
        "location_key",
        "date_key"
    ]
].head()

,provider_key,location_key,date_key
0,1,1,20240101
1,2,2,20240101
2,3,3,20240101
3,4,4,20240101
4,5,5,20240101


## Fact Table Construction

The fact table stores provider-level healthcare utilization, reimbursement, demographic, chronic condition, and risk metrics.

### Fact Grain

One record per provider per reporting year.

### Foreign Keys

- provider_key
- location_key
- date_key

### Measure Categories

- Utilization Measures
- Financial Measures
- Drug Measures
- Medical Measures
- Demographic Measures
- Chronic Condition Measures
- Risk Measures

In [45]:
measure_cols

['Tot_HCPCS_Cds',
 'Tot_Benes',
 'Tot_Srvcs',
 'Tot_Sbmtd_Chrg',
 'Tot_Mdcr_Alowd_Amt',
 'Tot_Mdcr_Pymt_Amt',
 'Tot_Mdcr_Stdzd_Amt',
 'Drug_Sprsn_Ind',
 'Drug_Tot_HCPCS_Cds',
 'Drug_Tot_Benes',
 'Drug_Tot_Srvcs',
 'Drug_Sbmtd_Chrg',
 'Drug_Mdcr_Alowd_Amt',
 'Drug_Mdcr_Pymt_Amt',
 'Drug_Mdcr_Stdzd_Amt',
 'Med_Sprsn_Ind',
 'Med_Tot_HCPCS_Cds',
 'Med_Tot_Benes',
 'Med_Tot_Srvcs',
 'Med_Sbmtd_Chrg',
 'Med_Mdcr_Alowd_Amt',
 'Med_Mdcr_Pymt_Amt',
 'Med_Mdcr_Stdzd_Amt',
 'Bene_Avg_Age',
 'Bene_Age_LT_65_Cnt',
 'Bene_Age_65_74_Cnt',
 'Bene_Age_75_84_Cnt',
 'Bene_Age_GT_84_Cnt',
 'Bene_Feml_Cnt',
 'Bene_Male_Cnt',
 'Bene_Race_Wht_Cnt',
 'Bene_Race_Black_Cnt',
 'Bene_Race_API_Cnt',
 'Bene_Race_Hspnc_Cnt',
 'Bene_Race_NatInd_Cnt',
 'Bene_Race_Othr_Cnt',
 'Bene_Dual_Cnt',
 'Bene_Ndual_Cnt',
 'Bene_CC_BH_ADHD_OthCD_V1_Pct',
 'Bene_CC_BH_Alcohol_Drug_V1_Pct',
 'Bene_CC_BH_Tobacco_V1_Pct',
 'Bene_CC_BH_Alz_NonAlzdem_V2_Pct',
 'Bene_CC_BH_Anxiety_V1_Pct',
 'Bene_CC_BH_Bipolar_V1_Pct',
 'Bene_CC_BH_Mood

## Fact Table Construction

Create the provider utilization fact table using surrogate keys and analytical measures.

The fact table serves as the central table within the healthcare analytics warehouse.

In [46]:
fact_provider_utilization = claims_fact[
    [
        "provider_key",
        "location_key",
        "date_key"
    ] + measure_cols
].copy()
fact_provider_utilization.shape

(1296739, 67)

In [47]:
print(f"Rows    : {fact_provider_utilization.shape[0]:,}")
print(f"Columns : {fact_provider_utilization.shape[1]:,}")

Rows    : 1,296,739
Columns : 67


In [48]:
fact_provider_utilization.head()

,provider_key,location_key,date_key,Tot_HCPCS_Cds,Tot_Benes,Tot_Srvcs,Tot_Sbmtd_Chrg,Tot_Mdcr_Alowd_Amt,Tot_Mdcr_Pymt_Amt,Tot_Mdcr_Stdzd_Amt,Drug_Sprsn_Ind,Drug_Tot_HCPCS_Cds,Drug_Tot_Benes,Drug_Tot_Srvcs,Drug_Sbmtd_Chrg,Drug_Mdcr_Alowd_Amt,Drug_Mdcr_Pymt_Amt,Drug_Mdcr_Stdzd_Amt,Med_Sprsn_Ind,Med_Tot_HCPCS_Cds,Med_Tot_Benes,Med_Tot_Srvcs,Med_Sbmtd_Chrg,Med_Mdcr_Alowd_Amt,Med_Mdcr_Pymt_Amt,Med_Mdcr_Stdzd_Amt,Bene_Avg_Age,Bene_Age_LT_65_Cnt,Bene_Age_65_74_Cnt,Bene_Age_75_84_Cnt,Bene_Age_GT_84_Cnt,Bene_Feml_Cnt,Bene_Male_Cnt,Bene_Race_Wht_Cnt,Bene_Race_Black_Cnt,Bene_Race_API_Cnt,Bene_Race_Hspnc_Cnt,Bene_Race_NatInd_Cnt,Bene_Race_Othr_Cnt,Bene_Dual_Cnt,Bene_Ndual_Cnt,Bene_CC_BH_ADHD_OthCD_V1_Pct,Bene_CC_BH_Alcohol_Drug_V1_Pct,Bene_CC_BH_Tobacco_V1_Pct,Bene_CC_BH_Alz_NonAlzdem_V2_Pct,Bene_CC_BH_Anxiety_V1_Pct,Bene_CC_BH_Bipolar_V1_Pct,Bene_CC_BH_Mood_V2_Pct,Bene_CC_BH_Depress_V1_Pct,Bene_CC_BH_PD_V1_Pct,Bene_CC_BH_PTSD_V1_Pct,Bene_CC_BH_Schizo_OthPsy_V1_Pct,Bene_CC_PH_Asthma_V2_Pct,Bene_CC_PH_Afib_V2_Pct,Bene_CC_PH_Cancer6_V2_Pct,Bene_CC_PH_CKD_V2_Pct,Bene_CC_PH_COPD_V2_Pct,Bene_CC_PH_Diabetes_V2_Pct,Bene_CC_PH_HF_NonIHD_V2_Pct,Bene_CC_PH_Hyperlipidemia_V2_Pct,Bene_CC_PH_Hypertension_V2_Pct,Bene_CC_PH_IschemicHeart_V2_Pct,Bene_CC_PH_Osteoporosis_V2_Pct,Bene_CC_PH_Parkinson_V2_Pct,Bene_CC_PH_Arthritis_V2_Pct,Bene_CC_PH_Stroke_TIA_V2_Pct,Bene_Avg_Risk_Scre
0,1,1,20240101,15,328,399.0,202783.88,46472.65,35325.46,34781.01,NaN,0.0,0.0,0.0,0.0,0.00,0.0,0.0,NaN,15.0,328.0,399.0,202783.88,46472.65,35325.46,34781.01,76,37.0,94.0,130.0,67.0,171.0,157.0,281.0,21.0,NaN,NaN,NaN,NaN,77.0,251.0,NaN,15.0,19.0,27.0,46.0,9.0,52.0,48.0,NaN,NaN,5.0,17.0,39.0,27.0,55.0,38.0,48.0,48.0,75.0,75.0,51.0,18.0,4.0,63.0,21.0,2.5660
1,2,2,20240101,21,912,2023.0,302581.00,69422.59,50178.11,47710.61,NaN,0.0,0.0,0.0,0.0,0.00,0.0,0.0,NaN,21.0,912.0,2023.0,302581.00,69422.59,50178.11,47710.61,76,16.0,419.0,337.0,140.0,455.0,457.0,806.0,13.0,29.0,23.0,0.0,41.0,48.0,864.0,NaN,3.0,4.0,5.0,16.0,3.0,20.0,17.0,1.0,NaN,NaN,6.0,17.0,17.0,21.0,10.0,23.0,14.0,75.0,64.0,26.0,15.0,2.0,47.0,6.0,1.1958
2,3,3,20240101,41,316,1369.0,381301.00,133096.84,101577.94,103554.75,NaN,2.0,20.0,261.0,4154.0,29.87,23.8,23.8,NaN,39.0,316.0,1108.0,377147.00,133066.97,101554.14,103530.95,70,84.0,127.0,84.0,21.0,192.0,124.0,199.0,97.0,NaN,NaN,0.0,NaN,112.0,204.0,NaN,20.0,21.0,7.0,53.0,7.0,46.0,44.0,NaN,4.0,5.0,15.0,17.0,12.0,34.0,32.0,45.0,27.0,75.0,75.0,36.0,17.0,NaN,75.0,11.0,1.6931
3,4,4,20240101,18,75,140.0,22015.00,8708.74,7134.32,7502.21,NaN,0.0,0.0,0.0,0.0,0.00,0.0,0.0,NaN,18.0,75.0,140.0,22015.00,8708.74,7134.32,7502.21,70,NaN,50.0,NaN,NaN,75.0,0.0,NaN,NaN,NaN,NaN,NaN,NaN,11.0,64.0,NaN,NaN,NaN,NaN,35.0,NaN,29.0,28.0,NaN,NaN,NaN,NaN,NaN,20.0,NaN,NaN,17.0,NaN,73.0,71.0,17.0,25.0,NaN,45.0,NaN,0.7931
4,5,5,20240101,28,101,136.0,120420.00,24416.29,18766.80,18618.81,NaN,0.0,0.0,0.0,0.0,0.00,0.0,0.0,NaN,28.0,101.0,136.0,120420.00,24416.29,18766.80,18618.81,68,NaN,49.0,25.0,NaN,47.0,54.0,74.0,NaN,0.0,11.0,0.0,NaN,34.0,67.0,NaN,13.0,NaN,NaN,33.0,NaN,35.0,30.0,NaN,NaN,NaN,16.0,NaN,16.0,23.0,16.0,25.0,13.0,56.0,55.0,26.0,NaN,NaN,54.0,NaN,1.4580


## Fact Table Standardization

Prepare analytical measures for warehouse loading by handling suppressed and missing values according to CMS reporting conventions.

### Rules

- Count measures → replace null values with 0
- Financial measures → replace null values with 0
- Percentage metrics → preserve null values
- Risk metrics → preserve null values

In [49]:
fact_provider_utilization.isnull() \
    .sum() \
    .sort_values(ascending=False) \
    .head(30)

Med_Sprsn_Ind                      1145043
Drug_Sprsn_Ind                     1145043
Bene_Race_API_Cnt                  1062540
Bene_Race_Othr_Cnt                 1052600
Bene_Race_Hspnc_Cnt                 988886
Bene_Race_Black_Cnt                 924899
Bene_CC_BH_PTSD_V1_Pct              882169
Bene_CC_BH_ADHD_OthCD_V1_Pct        875024
Bene_CC_BH_PD_V1_Pct                852076
Bene_Race_NatInd_Cnt                848880
Bene_CC_PH_Parkinson_V2_Pct         804287
Bene_CC_BH_Schizo_OthPsy_V1_Pct     767990
Bene_CC_BH_Bipolar_V1_Pct           735331
Bene_Age_LT_65_Cnt                  690801
Bene_Age_GT_84_Cnt                  678412
Bene_CC_BH_Alcohol_Drug_V1_Pct      627335
Bene_CC_BH_Alz_NonAlzdem_V2_Pct     559099
Bene_CC_PH_Stroke_TIA_V2_Pct        529707
Bene_CC_BH_Tobacco_V1_Pct           522349
Bene_CC_PH_Asthma_V2_Pct            481345
Bene_Ndual_Cnt                      464482
Bene_Dual_Cnt                       464482
Bene_Race_Wht_Cnt                   462752
Bene_CC_PH_

## Warehouse Null Handling Strategy

The CMS dataset contains suppressed values used to protect beneficiary privacy.

### Count Measures

Null values are replaced with 0.

### Financial Measures

Null values are replaced with 0.

### Percentage Measures

Null values are preserved.

### Risk Measures

Null values are preserved.

### Suppression Indicators

Null values are preserved.

In [50]:
count_measure_cols = [
    "Drug_Tot_HCPCS_Cds",
    "Drug_Tot_Benes",
    "Drug_Tot_Srvcs",
    "Med_Tot_HCPCS_Cds",
    "Med_Tot_Benes",
    "Med_Tot_Srvcs",
    "Bene_Age_LT_65_Cnt",
    "Bene_Age_65_74_Cnt",
    "Bene_Age_75_84_Cnt",
    "Bene_Age_GT_84_Cnt",
    "Bene_Feml_Cnt",
    "Bene_Male_Cnt",
    "Bene_Race_Wht_Cnt",
    "Bene_Race_Black_Cnt",
    "Bene_Race_API_Cnt",
    "Bene_Race_Hspnc_Cnt",
    "Bene_Race_NatInd_Cnt",
    "Bene_Race_Othr_Cnt",
    "Bene_Dual_Cnt",
    "Bene_Ndual_Cnt"
]
fact_provider_utilization[
    count_measure_cols
].isnull().sum().sum()

np.int64(9517130)

## Count Measure Standardization

Replace null values in additive count measures with zero.

This ensures accurate aggregation in SQL, PostgreSQL, and Power BI while preserving suppression behavior in percentage-based metrics.

In [51]:
fact_provider_utilization[count_measure_cols] = (
    fact_provider_utilization[count_measure_cols]
    .fillna(0)
)

In [52]:
fact_provider_utilization[
    count_measure_cols
].isnull().sum().sum()

np.int64(0)

In [53]:
fact_provider_utilization[
    [
        "Bene_Race_API_Cnt",
        "Bene_Dual_Cnt",
        "Bene_Age_LT_65_Cnt"
    ]
].head()

,Bene_Race_API_Cnt,Bene_Dual_Cnt,Bene_Age_LT_65_Cnt
0,0.0,77.0,37.0
1,29.0,48.0,16.0
2,0.0,112.0,84.0
3,0.0,11.0,0.0
4,0.0,34.0,0.0


## Warehouse Validation

Validate dimension and fact table integrity before loading into the warehouse layer.

Checks:

- Dimension uniqueness
- Foreign key completeness
- Fact table row count
- Null key validation

In [54]:
print("=" * 50)
print("WAREHOUSE VALIDATION")
print("=" * 50)
print(f"Provider Dimension Rows : {len(dim_provider):,}")
print(f"Location Dimension Rows : {len(dim_location):,}")
print(f"Date Dimension Rows     : {len(dim_date):,}")
print()
print(f"Fact Table Rows         : {len(fact_provider_utilization):,}")
print()
print(
    f"Null Provider Keys      : "
    f"{fact_provider_utilization['provider_key'].isnull().sum():,}"
)
print(
    f"Null Location Keys      : "
    f"{fact_provider_utilization['location_key'].isnull().sum():,}"
)
print(
    f"Null Date Keys          : "
    f"{fact_provider_utilization['date_key'].isnull().sum():,}"
)

WAREHOUSE VALIDATION
Provider Dimension Rows : 1,296,739
Location Dimension Rows : 26,836
Date Dimension Rows     : 1

Fact Table Rows         : 1,296,739

Null Provider Keys      : 0
Null Location Keys      : 0
Null Date Keys          : 0


In [57]:
fact_provider_utilization[
    [
        "provider_key",
        "location_key",
        "date_key"
    ]
].duplicated().sum()

np.int64(0)

# Warehouse Layer Export

Persist dimension and fact tables into the warehouse layer.

These files represent the curated analytical model that will be loaded into PostgreSQL and consumed by reporting tools.

In [58]:
from pathlib import Path
warehouse_path = Path("../warehouse")
warehouse_path.mkdir(
    parents=True,
    exist_ok=True
)
print("Warehouse directory ready.")

Warehouse directory ready.


In [59]:
dim_provider.to_csv("../warehouse/dim_provider.csv", index=False)
dim_location.to_csv("../warehouse/dim_location.csv", index=False)
dim_date.to_csv("../warehouse/dim_date.csv", index=False)
print("Dimension exports complete.")

Dimension exports complete.


In [60]:
fact_provider_utilization.to_csv("../warehouse/fact_provider_utilization.csv", index=False)
print("Fact table export complete.")

Fact table export complete.


In [61]:
import os
for file in sorted(os.listdir("../warehouse")):
    print(file)

dim_date.csv
dim_location.csv
dim_provider.csv
fact_provider_utilization.csv


In [66]:
fact_provider_utilization.columns.tolist()

['provider_key',
 'location_key',
 'date_key',
 'Tot_HCPCS_Cds',
 'Tot_Benes',
 'Tot_Srvcs',
 'Tot_Sbmtd_Chrg',
 'Tot_Mdcr_Alowd_Amt',
 'Tot_Mdcr_Pymt_Amt',
 'Tot_Mdcr_Stdzd_Amt',
 'Drug_Sprsn_Ind',
 'Drug_Tot_HCPCS_Cds',
 'Drug_Tot_Benes',
 'Drug_Tot_Srvcs',
 'Drug_Sbmtd_Chrg',
 'Drug_Mdcr_Alowd_Amt',
 'Drug_Mdcr_Pymt_Amt',
 'Drug_Mdcr_Stdzd_Amt',
 'Med_Sprsn_Ind',
 'Med_Tot_HCPCS_Cds',
 'Med_Tot_Benes',
 'Med_Tot_Srvcs',
 'Med_Sbmtd_Chrg',
 'Med_Mdcr_Alowd_Amt',
 'Med_Mdcr_Pymt_Amt',
 'Med_Mdcr_Stdzd_Amt',
 'Bene_Avg_Age',
 'Bene_Age_LT_65_Cnt',
 'Bene_Age_65_74_Cnt',
 'Bene_Age_75_84_Cnt',
 'Bene_Age_GT_84_Cnt',
 'Bene_Feml_Cnt',
 'Bene_Male_Cnt',
 'Bene_Race_Wht_Cnt',
 'Bene_Race_Black_Cnt',
 'Bene_Race_API_Cnt',
 'Bene_Race_Hspnc_Cnt',
 'Bene_Race_NatInd_Cnt',
 'Bene_Race_Othr_Cnt',
 'Bene_Dual_Cnt',
 'Bene_Ndual_Cnt',
 'Bene_CC_BH_ADHD_OthCD_V1_Pct',
 'Bene_CC_BH_Alcohol_Drug_V1_Pct',
 'Bene_CC_BH_Tobacco_V1_Pct',
 'Bene_CC_BH_Alz_NonAlzdem_V2_Pct',
 'Bene_CC_BH_Anxiety_V1_Pct',
